pallets/pallet-bonded-coins/src/lib.rs

```diff
@@ -199,32 +227,274 @@ pub mod pallet {
+				return Ok(());
+			}
+
+			let transferred = T::CollateralCurrency::transfer(
+				T::CollateralAssetId::get(),
+				&pool_account,
+				&who,
+				amount,
+				Preservation::Expendable,
+			)?; // TODO: check edge cases around existential deposit
+
+			// if collateral or total supply drops to zero, refunding is complete -> emit event
+			if sum_of_issuances <= burnt || total_collateral_issuance <= transferred {
+				Self::deposit_event(Event::RefundComplete { id: pool_id });
+			}
+
+			Ok(())
 		}
 
 		#[pallet::call_index(9)]
 		#[pallet::weight(Weight::from_parts(10_000, 0) + T::DbWeight::get().writes(1))]
-		pub fn finish_destroy(_origin: OriginFor<T>) -> DispatchResult {
-			todo!()
+		pub fn start_destroy(origin: OriginFor<T>, pool_id: T::PoolId, currency_count: u32) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			Self::do_start_destroy_pool(pool_id, Some(&who), false, currency_count)?;
+
+			Ok(())
+		}
+
+		#[pallet::call_index(10)]
+		#[pallet::weight(Weight::from_parts(10_000, 0) + T::DbWeight::get().writes(1))]
+		pub fn force_start_destroy(origin: OriginFor<T>, pool_id: T::PoolId, currency_count: u32) -> DispatchResult {
+			T::ForceOrigin::ensure_origin(origin)?;
+
+			Self::do_start_destroy_pool(pool_id, None, true, currency_count)?;
+
+			Ok(())
+		}
+
+		#[pallet::call_index(11)]
+		#[pallet::weight(Weight::from_parts(10_000, 0) + T::DbWeight::get().writes(1))]
+		pub fn finish_destroy(origin: OriginFor<T>, pool_id: T::PoolId, currency_count: u32) -> DispatchResult {
+			ensure_signed(origin)?;
+
+			let pool_details = Pools::<T>::get(&pool_id).ok_or(Error::<T>::PoolUnknown)?;
+
+			ensure!(
+				pool_details.bonded_currencies.len() <= currency_count.saturated_into(),
+				Error::<T>::TooManyCurrencies
+			);
+
+			ensure!(pool_details.state.is_destroying(), Error::<T>::LivePool);
+
+			for asset_id in pool_details.bonded_currencies {
+				if T::Fungibles::asset_exists(asset_id.clone()) {
+					// This would fail with an LiveAsset error if there are any accounts left on any currency
+					T::Fungibles::finish_destroy(asset_id)?;
+				}
+			}
+
+			let pool_account = pool_id.clone().into();
+
+			let total_collateral_issuance = Self::get_pool_collateral(&pool_account);
+
+			if !total_collateral_issuance.is_zero() {
+				T::CollateralCurrency::transfer(
+					T::CollateralAssetId::get(),
+					&pool_account,
+					&pool_details.manager,
+					total_collateral_issuance,
+					Preservation::Expendable,
+				)?;
+			}
+
+			Pools::<T>::remove(&pool_id);
+
+			// TODO: refund deposit
+
+			Self::deposit_event(Event::Destroyed { id: pool_id });
+
+			Ok(())
 		}
 	}
 
-	impl<T: Config> Pallet<T> {}
+	impl<T: Config> Pallet<T> {
+		fn do_start_refund(
+			pool_id: T::PoolId,
+			maybe_check_manager: Option<&AccountIdOf<T>>,
+			max_currencies: u32,
+		) -> Result<u32, DispatchError> {
+			let mut pool_details = Pools::<T>::get(&pool_id).ok_or(Error::<T>::PoolUnknown)?;
```

Comment:
One more trick you could use, is to declare it here as non mut, and redeclare it as `mut` before you actually modify its state, so you avoid any unwanted side-effects that could slip in.

Reply:
Even better, in my opinion, if you don't have it as mut at all, and simply call `Pools::<T>::mutate` and change whatever needs changing in there. But this is more of a nit than the previous comment above, which is more of a recommendation.

Reply:
The selective application of mut is a smart move, will do that! Calling `get` and `mutate` in the same function has a weird smell for me though - having to grab the same record twice, including handling not-found error cases twice, seems a bit like a red flag to me.

Reply:
Also `mutate` does not accept failing, as opposed to `try_mutate` which is for cases where you want to check and then mutate.

Reply:
I don't think you don't need to clone since you're simply re-assigning the ownership. If you're getting a compilation error without cloning, please resolve this conversation yourself.

Reply:
I did get compilation errors since I accessed the bonded_currencies later, but this solves it: d38e602ca5826843403f58a72ec74b2d5021c1ff